### PR TITLE
Implement asm-parser for dotnet

### LINKED ISSUE #3325: C# code is not shown as optimized
https://github.com/compiler-explorer/compiler-explorer/pull/3168 added support for C#, but as it looks the output isn't optimized and looks more like a debug-output.
Different compiler-flags as listed in the PR's description and some comments don't affect the output neither.

E.g. for https://godbolt.org/z/qdqdj5ffY expected output should be similar to
```asm
Foo:Bar(System.Span`1[System.Byte],long):ubyte:
    mov rax, [rcx]
    movzx eax, byte ptr [rax+rdx]
    ret
```
as for instance [Sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8YUCAAcAymIBuvMDFwBuGvSatOPfoIDCEfBN4AbMTKjzFKmmtIMAYhAg0A3jQYemAZg1IGwAE8MGAYAIWwoAAopCWwuAB5A4IA+BlxYrjQGLl4eBlyAExgEAEp3TzdqT2qGWAAzfyCQqGAGAF5amAaAWRh8aADuiNwAC2xjFgBxGAw2LrEYLkVI9LiS1Sqaj2IAdgYAVS5cbDrBAEECgsj62uAswuL18o8AXxoXoA===) shows.

### PR DESCRIPTION
**Pull request by hez2010**
- Implement asm-parser for dotnet.
- Update the README.md to reflect language changes.
- Fix codegen perf issue while compiling code against intrinsics.
- Fix issue where there's an empty line after method signature.

Before/After unused labels filter:

image

image

Fixes #3325

### CODE REVIEW

**Comment by partouf** [comment]
This is not the best way to do this. The best way would be to make a custom asm parser that can deal with these cases. So instead of having the cleanAsm function recognize the start of functions, you let the parser do that.

**Comment by partouf** [comment]
Perhaps the best place if any to ask this; do you know of any command line parameters for crossgen2 to include line number annotations in the assembly?
How much noise does `-g` give? https://github.com/dotnet/runtimelab/blob/d0a37893a67c125f9b0cd8671846ff7d867df241/src/coreclr/src/tools/aot/ILCompiler/Program.cs#L160
And if that's not useful, how much info does the Map file give, can we tell it to include linenumbers? https://github.com/dotnet/runtimelab/blob/d0a37893a67c125f9b0cd8671846ff7d867df241/src/coreclr/src/tools/aot/ILCompiler/Program.cs#L174

**Comment by EgorBo** [comment]
> Perhaps the best place if any to ask this; do you know of any command line parameters for crossgen2 to include line number annotations in the assembly?
> 
> Looks like the mapfile option cannot help us here, I don't see any structures in the dotnet code to suggest that they store sourcefile and linenumber references for symbols.

I was planning some changes for crossgen2 for godbolt specially, how the line numbers should look like?

**Comment by partouf** [comment]
> > Perhaps the best place if any to ask this; do you know of any command line parameters for crossgen2 to include line number annotations in the assembly?
> > Looks like the mapfile option cannot help us here, I don't see any structures in the dotnet code to suggest that they store sourcefile and linenumber references for symbols.
> 
> I was planning some changes for crossgen2 for godbolt specially, how the line numbers should look like?

With a custom `asm-parser.js` you can basically do whatever you think is easiest.

If you want to reuse an existing technique, the best supported is:
* `.file fileindex filepath` - once per file anywhere in the assembly
* `.loc fileindex linenumber [columnnumber]` before assembly thats about that file and linenumber, it's usually placed after the function label (don't ask why, I wish it were different)

**Comment by ShreyasJejurkar** [comment]
Along with this, is there any way we can fix this formatting of assembly code? Like in the below screenshot, can we get rid of these empty lines? Line number 2 

image

**Comment by RubenRBS** [comment]
I think it's this newline at the end of the string https://github.com/compiler-explorer/compiler-explorer/pull/3334/files#diff-f47b98865b3a9b0d5178e6b55305ffc1c8cf591fb7424d121311938951d3b579L171 right?

**Comment by ShreyasJejurkar** [comment]
@RubenRBS yeah, maybe. Not sure about the code which is causing that empty line!
Just pointed out here, in case if we can fix that too.
Am pretty new to this compiler explore codebase entirely, so cannot pin point it!

File: README.md [hunk 1]
```diff
@@ -5,8 +5,8 @@
 
 # Compiler Explorer
 
-**Compiler Explorer** is an interactive compiler exploration website. Edit code in C, C++, Rust, Go, D, Haskell, Swift, Pascal, [ispc](https://ispc.github.io/), Python, Java
- or in any of the other [31 supported languages](https://godbolt.org/api/languages), and see how that code looks after being compiled in real time.
+**Compiler Explorer** is an interactive compiler exploration website. Edit code in C, C++, C#, F#, Rust, Go, D, Haskell, Swift, Pascal, [ispc](https://ispc.github.io/), Python, Java
+ or in any of the other [35 supported languages](https://godbolt.org/api/languages), and see how that code looks after being compiled in real time.
```
Comment (mattgodbolt) [inline README.md:9]: We should probably replace `35` with `30+` to cover the future :)

File: lib/compilers/dotnet.ts [hunk 2]
```diff
@@ -168,7 +168,7 @@ class DotNetCompiler extends BaseCompiler {
                 // ; Assembly listing for method ConsoleApplication.Program:Main(System.String[])
                 //                               ^ This character is the 31st character in this string.
                 // `substring` removes the first 30 characters from it and uses the rest as a label.
-                cleanedAsm = cleanedAsm.concat(line.text.substring(30) + ':\n');
+                cleanedAsm = cleanedAsm.concat('Method ' + line.text.substring(30) + ':\n');
```
Comment (mattgodbolt) [inline lib/compilers/dotnet.ts:171]: +1 to @partouf 'sthoughts that there must be a better way in general than this. Let us know if we can help with direction.

Comment (hez2010) [reply lib/compilers/dotnet.ts:171]: I'm working on this. Will push a commit later.

Comment (mattgodbolt) [reply lib/compilers/dotnet.ts:171]: Thanks so much: do let us know when it's ready for another look!

**Comment by hez2010** [comment]
@partouf @mattgodbolt PTAL. I've implemented the asm-parser for dotnet.

**Comment by partouf** [comment]
@hez2010 looks great, thanks!

**Comment by partouf** [comment]
TODO @partouf refactor and unify typescript types with https://github.com/compiler-explorer/compiler-explorer/pull/3278/files#diff-8e2abda8dacea4bfc00bab8bc7b8bc4c41ffe114a92dedd634e4071268652df7

**Comment by hez2010** [comment]
@partouf Do you mind if I include a commit which can fix the performance issue in #3325?

**Comment by partouf** [comment]
Sure, go ahead @hez2010

**Comment by hez2010** [comment]
Done.

File: lib/compilers/dotnet.ts (hunk 2 above)
Comment (hez2010) [reply lib/compilers/dotnet.ts]: Done.

File: lib/compilers/dotnet.ts [hunk 3]
```diff
@@ -160,36 +163,11 @@ class DotNetCompiler extends BaseCompiler {
...
-        }
-
-        return cleanedAsm;
-    }
-
     async runCrossgen2(compiler, execOptions, crossgen2Path, publishPath, dllPath, options, outputPath) {
         const crossgen2Options = [
             crossgen2Path, '-r', path.join(publishPath, '*'), dllPath, '-o', 'CompilerExplorer.r2r.dll',
             '--codegenopt', 'NgenDisasm=*', '--codegenopt', 'JitDiffableDasm=1', '--parallelism', '1',
+            '--inputbubble', '--compilebubblegenerics', '--resilient',
```
Comment (EgorBo) [inline lib/compilers/dotnet.ts:170]: please remove --resilient here, it will hide potential errors we don't want to see at all

Comment (hez2010) [reply lib/compilers/dotnet.ts]: Done.

**Comment by hez2010** [comment]
Oops.. Just found that forgot to include asm-parser-dotnet in the GitHub labeler: https://github.com/compiler-explorer/compiler-explorer/blob/f778e769cb85e0c60867c8b366572d9def3e007e/.github/labeler.yml#L45

**Comment by partouf** [comment]
Done

**Comment by partouf** [comment]
this is lvei

**Pull request merged**